### PR TITLE
Improved request interception semantics and `HttpAdapter` configuration performance

### PR DESCRIPTION
**Pull request by kyri-petrou**
This PR:

1. Adds an `interceptWithZIO` method on `HttpAdapter` which can be used to intercept requests and apply some logic without erasing part of the environment.
2. Improves performance for configured `HttpAdapter`s by avoiding modification of the ZIO environment (see [Discord post](https://discord.com/channels/629491597070827530/633200096393166868/1133309512284114965) for more info)

⚠️ Bincompat Breaking change ⚠️

To provide a unified `configure` implementation, the method now requires an implicit `Tag[R1]`. This is not necessary if we wish to have different `configure` implementations across `Base` and `Intercepted`. Let me know whether you think it's better to avoid the breaking change.

On this note, I think we might want to also add an implicit `Tag[R1]` to `intercept` since it seems that different layer composition methods require one, which we might need to use in the future

## Todo

- [ ] Add tests for `interceptWithZIO`

### CODE REVIEW

File: interop/tapir/src/main/scala/caliban/interop/tapir/HttpInterpreter.scala [hunk 1]
```diff
@@ -52,15 +64,43 @@ object HttpInterpreter {
...
+      interceptor: ServerRequest => ZIO[R2, TapirResponse, Unit]
+    )(implicit tag: Tag[R2]): HttpInterpreter[R1 & R2, E] =
+      copy[R1 & R2, R & R2, E](layer =
+        layer ++ ZLayer.environment[R2] ++ ZLayer(ZIO.serviceWithZIO[ServerRequest](interceptor))
+      )
+
+    def configure[R2](configurator: Configurator[R2])(implicit tag: Tag[R2]): HttpInterpreter[R1 & R2, E] =
+      Intercepted[R1 & R2, R & R2, E](
+        interpreter.configure[R2](configurator),
+        layer ++ ZLayer.environment[R2]
```
Comment (ghostdogpr) [inline interop/tapir/src/main/scala/caliban/interop/tapir/HttpInterpreter.scala:110]: Using `ZLayer.makeSome` doesn't require a `Tag` and is faster, I think?

File: interop/tapir/src/main/scala/caliban/interop/tapir/HttpInterpreter.scala [hunk 2]
```diff
@@ -52,15 +64,43 @@ object HttpInterpreter {
       graphQLRequest: GraphQLRequest,
       serverRequest: ServerRequest
     )(implicit streamConstructor: StreamConstructor[BS]): ZIO[R, TapirResponse, CalibanResponse[BS]] =
-      interpreter.executeRequest(graphQLRequest).map(buildHttpResponse[E, BS])
+      ZIO.scoped[R](config *> interpreter.executeRequest(graphQLRequest).map(buildHttpResponse[E, BS]))
```
Comment (ghostdogpr) [inline interop/tapir/src/main/scala/caliban/interop/tapir/HttpInterpreter.scala]: Seems pity to add overhead on the simple case.

File: interop/tapir/src/main/scala/caliban/interop/tapir/HttpInterpreter.scala (hunk 1 above)
Comment (kyri-petrou) [reply interop/tapir/src/main/scala/caliban/interop/tapir/HttpInterpreter.scala:110]: I’ve tried everything I could think of to get rid of the `Tag` / `++` and I can’t seem to get it to work (same for `configure`). 

When using `ZLayer.makeSome` alongside `ZLayer.environment`, a circular dependency compilation error is raised if the layer output contains a union type.

**Comment by ghostdogpr** [comment]
I'm not so keen on making the API more complex 🤔 How about something like this instead? https://github.com/ghostdogpr/caliban/pull/1816

**Pull request closed without merging**